File: Java/src/Data/DT_Data.java
package Data;

import DataObserver.*;


import java.util.ArrayList;
import java.util.List;



/**
 * Abstract class representing generic data in the Digital Twin framework.
 * This class serves as a base for different specialized data types (e.g., {@code SystemData}, {@code PlatformData}),
 * providing shared attributes and methods for data management.
 *
 * <p>Key features of the class include:
 * <ul>
 *   <li>Encapsulation of common data attributes such as source, data type, and timestamp.</li>
 *   <li>Abstract methods for fetching and validating data to enforce implementation in derived classes.</li>
 *   <li>A standard {@code toString} method for debugging and logging purposes.</li>
 * </ul>
 * </p>
 */
public abstract class DT_Data<T> implements ObserverDataManager<T> {

    // List to hold all observers interested in this data type
    private final List<ObserverData<T>> observers = new ArrayList<>();

    // Holds the latest data
    private T currentData;

    @Override
    public void registerObserver(ObserverData<T> observer) {
        if (!observers.contains(observer)) {
            observers.add(observer);
            System.out.println("Observer registered: " + observer.getClass().getSimpleName());
        }
    }

    @Override
    public void removeObserver(ObserverData<T> observer) {
        observers.remove(observer);
        System.out.println("Observer removed: " + observer.getClass().getSimpleName());
    }

    @Override
    public void notifyObservers() {
        // Safe copy to avoid issues if observer removes itself during update
        List<ObserverData<T>> observersCopy = new ArrayList<>(observers);

        for (ObserverData<T> observer : observersCopy) {
            observer.update(currentData);
        }
    }

    /**
     * Call this method whenever new data arrives
     */
    public void updateData(T newData) {
        this.currentData = newData;
        System.out.println("\n[DATA SOURCE] New data received: " + newData);
        notifyObservers();
    }

    /**
     * Get the latest data
     */
    public T getCurrentData() {
        return currentData;
    }



/*
private final List<ObserverData<T>> observers = new ArrayList<>();
    private int temperature;           // Example data field (you can add more)
    private String location = "City Center"; // Another field for realism

    // === Subject methods (from DataSubject) ===
    @Override
    public void registerObserver(ObserverData<DT_Data> observer) {
        observers.add(observer);
        System.out.println("Observer registered: " + observer.getClass().getSimpleName());
    }

    @Override
    public void removeObserver(ObserverData<DT_Data> observer) {
        observers.remove(observer);
        System.out.println("Observer removed: " + observer.getClass().getSimpleName());
    }

    @Override
    public void notifyObservers() {
        // Safe copy to avoid ConcurrentModificationException if an observer removes itself
        List<ObserverData<DT_Data>> observersCopy = new ArrayList<>(observers);
        for (ObserverData<DT_Data> observer : observersCopy) {
            observer.update(this);
        }
    }

    // === Your data-update method (this is where "new data comes") ===
    public void setNewTemperature(int newTemperature) {
        this.temperature = newTemperature;
        System.out.println("\n[DATA] New temperature received: " + newTemperature + "°C");
        notifyObservers();   // <-- This automatically updates all Models & Applications
    }

    // Getters so observers can read the latest data
    public int getTemperature() {
        return temperature;
    }

    public String getLocation() {
        return location;
    }
    */





    /**
     * The source of the data (e.g., "Sensor", "API",Human,...).
     */
    protected String source;

    /**
     * The type of the data (e.g., "Temperature", "Pressure").
     * */
    protected String dataType;

    /**
     * The timestamp indicating when the data was collected.
     * */
    protected String timestamp;

    /**
     * Constructs a new {@code DigitalTwin_DMS_Framework.Data} object with the specified attributes.
     *
     * @param source     The origin of the data (e.g., sensor name or system name).
     * @param dataType   The type or category of the data (e.g., temperature, text).
     * @param timestamp  The time at which the data was recorded, in an ISO-compliant format.
     */
    public DT_Data(String source, String dataType, String timestamp) {
        this.source = source;
        this.dataType = dataType;
        this.timestamp = timestamp;
    }

    /**
     * Fetches data from the source.
     * Subclasses must implement specific logic for fetching data based on their context.
     */
    public abstract void fetchData();

    /**
     * Validates the integrity and completeness of the data.
     * Subclasses must define their own validation criteria and logic.
     *
     * @return {@code true} if the data is valid; {@code false} otherwise.
     */
    /**
     * Validates the common attributes of the data object.
     * Subclasses can extend this validation as needed.
     *
     * @return true if the basic data attributes are valid; false otherwise.
     */
    public boolean validateData() {
        return source != null && !source.isEmpty() &&
                dataType != null && !dataType.isEmpty() &&
                timestamp != null && !timestamp.isEmpty();
    }

    /**
     * Retrieves the source of the data.
     *
     * @return The source of the data.
     */
    public String getSource() {
        return source;
    }

    /**
     * Sets a new source for the data.
     *
     * @param source The new source of the data.
     */
    public void setSource(String source) {
        this.source = source;
    }

    /**
     * Retrieves the type of the data.
     *
     * @return The data type.
     */
    public String getDataType() {
        return dataType;
    }

    /**
     * Sets a new type for the data.
     *
     * @param dataType The new data type.
     */
    public void setDataType(String dataType) {
        this.dataType = dataType;
    }

    /**
     * Retrieves the timestamp of the data.
     *
     * @return The timestamp when the data was collected.
     */
    public String getTimestamp() {
        return timestamp;
    }

    /**
     * Sets a new timestamp for the data.
     *
     * @param timestamp The new timestamp.
     */
    public void setTimestamp(String timestamp) {
        this.timestamp = timestamp;
    }

    /**
     * Converts the {@code DigitalTwin_DMS_Framework.Data} object into a human-readable string format.
     *
     * @return A string representation of the data object, including its attributes.
     */
    @Override
    public String toString() {
        return "DigitalTwin_DMS_Framework.Data{" +
                "source='" + source + '\'' +
                ", dataType='" + dataType + '\'' +
                ", timestamp='" + timestamp + '\'' +
                '}';
    }

    /**
     * Compares this {@code DigitalTwin_DMS_Framework.Data} object with another object for equality.
     * Two {@code DigitalTwin_DMS_Framework.Data} objects are considered equal if their attributes match.
     *
     * @param obj The object to compare with this instance.
     * @return {@code true} if the objects are equal; {@code false} otherwise.
     */
    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (!(obj instanceof DT_Data)) return false;

        if (!(obj instanceof DT_Data<?>)) 
        {
            return false;
        }

        DT_Data<?> data = (DT_Data<?>) obj;

        return source.equals(data.source)
            && dataType.equals(data.dataType)
            && timestamp.equals(data.timestamp);
            
    }

    /**
     * Computes the hash code for this {@code DigitalTwin_DMS_Framework.Data} object.
     * The hash code is based on the source, data type, and timestamp.
     *
     * @return The computed hash code.
     */
    @Override
    public int hashCode() {
        int result = source.hashCode();
        result = 31 * result + dataType.hashCode();
        result = 31 * result + timestamp.hashCode();
        return result;
    }
}

File: Python/src/DT_Data.py
from abc import ABC, abstractmethod
from typing import Generic, TypeVar, List

T = TypeVar('T')


class ObserverData(ABC, Generic[T]):
    @abstractmethod
    def update(self, data: T):
        pass


class ObserverDataManager(ABC, Generic[T]):
    @abstractmethod
    def register_observer(self, observer: ObserverData[T]):
        pass

    @abstractmethod
    def remove_observer(self, observer: ObserverData[T]):
        pass

    @abstractmethod
    def notify_observers(self):
        pass


class DT_Data(ObserverDataManager[T], ABC, Generic[T]):

    def __init__(self, source: str, data_type: str, timestamp: str):
        self._observers: List[ObserverData[T]] = []
        self._current_data: T | None = None

        self.source = source
        self.data_type = data_type
        self.timestamp = timestamp

    # === Observer Management ===

    def register_observer(self, observer: ObserverData[T]):
        if observer not in self._observers:
            self._observers.append(observer)
            print(f"Observer registered: {observer.__class__.__name__}")

    def remove_observer(self, observer: ObserverData[T]):
        if observer in self._observers:
            self._observers.remove(observer)
            print(f"Observer removed: {observer.__class__.__name__}")

    def notify_observers(self):
        # Safe copy to avoid modification during iteration
        for observer in list(self._observers):
            observer.update(self._current_data)

    # === Data Handling ===

    def update_data(self, new_data: T):
        self._current_data = new_data
        print(f"\n[DATA SOURCE] New data received: {new_data}")
        self.notify_observers()

    def get_current_data(self) -> T | None:
        return self._current_data

    # === Abstract Method ===

    @abstractmethod
    def fetch_data(self):
        pass

    # === Validation ===

    def validate_data(self) -> bool:
        return all([
            self.source,
            self.data_type,
            self.timestamp
        ])

    # === Getters / Setters ===

    def get_source(self) -> str:
        return self.source

    def set_source(self, source: str):
        self.source = source

    def get_data_type(self) -> str:
        return self.data_type

    def set_data_type(self, data_type: str):
        self.data_type = data_type

    def get_timestamp(self) -> str:
        return self.timestamp

    def set_timestamp(self, timestamp: str):
        self.timestamp = timestamp

    # === Object Overrides ===

    def __str__(self):
        return (
            f"DT_Data(source='{self.source}', "
            f"data_type='{self.data_type}', "
            f"timestamp='{self.timestamp}')"
        )

    def __eq__(self, other):
        if not isinstance(other, DT_Data):
            return False
        return (
            self.source == other.source and
            self.data_type == other.data_type and
            self.timestamp == other.timestamp
        )

    def __hash__(self):
        return hash((self.source, self.data_type, self.timestamp))
